Modernize ESProducers in JetMETCorrections/Modules

makortel opened the pull request:
#### PR description:

This PR modernizes ESProducers in JetMETCorrections/Modules (as part of #26748) by
* Fixing a memory leak in `JetCorrectionESSource`
* Using ESGetToken
* Merging header and source for QGLikelihoodESProducer, QGLikelihoodSystematicsESProducer, and JetCorrectionESChain, and moving them to `plugins`
* Avoiding a copy of the product in QGLikelihoodESProducer and QGLikelihoodSystematicsESProducer
* Using `make_unique`, and removing unnecessary call to `ChainedJetCorrector::clear()` in JetCorrectionESChain

#### PR validation:

Code compiles.

makortel committed:
Fix a memory leak in JetCorrectionESSource.h

makortel committed:
Use ESGetToken in JetCorrectionESProducer and JetResolutionESProducer

makortel committed:
Modernize QGLikelihoodESProducer

- Merge source and header and move to plugins
- Use ESGetToken

makortel committed:
Modernize QGLikelihoodSystematicsESProducer

- Merge source and header, and move to plugins
- Use ESGetToken

makortel committed:
Reduce copying in QGLikelihood(Systematics)ESProducer

makortel committed:
Modernize JetCorrectionESChain

- Merge source and header, and move to plugins
- Use ESGetToken
- Use make_unique
- Remove unnecessary call to ChainedJetCorrector::clear()

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel commented on JetMETCorrections/Modules/plugins/QGLikelihoodESProducer.cc:
```diff
@@ -0,0 +1,58 @@
+  ~QGLikelihoodESProducer() override{};
+
+  std::shared_ptr<const QGLikelihoodObject> produce(const QGLikelihoodRcd &);
+  void setIntervalFor(const edm::eventsetup::EventSetupRecordKey &, const edm::IOVSyncValue &, edm::ValidityInterval &);
```
@Dr15Jones Does the `setIntervalFor()` acqually do anything when the ESProducer does not inherit from `edm::EventSetupRecordIntervalFinder`?

Dr15Jones replied:
It won’t be called.

makortel replied:
Thanks, I'll remove those

makortel commented on JetMETCorrections/Modules/plugins/QGLikelihoodESProducer.cc:
```diff
@@ -0,0 +1,58 @@
+
+// Produce the data
+std::shared_ptr<const QGLikelihoodObject> QGLikelihoodESProducer::produce(const QGLikelihoodRcd& iRecord) {
+  return std::shared_ptr<const QGLikelihoodObject>(&iRecord.get(token_), edm::do_nothing_deleter());
```
Is the point of this class and QGLikelihoodSystematicsESProducer to act as an alias for the `QGLikelihood(Systematics)Object`? (i.e. provide a constant product label for the clients, while being able to have a single point where to make changes in case the product needs to point to something else in the DB(?))

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27216/10387

- This PR adds an extra 20KB to repository

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

JetMETCorrections/Modules


@perrotta, @cmsbuild, @santocch, @slava77 can you please review it and eventually sign? Thanks.
@TaiSakuma, @gouskos, @ahinzmann, @rappoccio, @mmarionncern, @jdamgov, @jdolen, @nhanvtran, @gkasieczka, @clelange, @schoef, @mariadalfonso, @seemasharmafnal this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

makortel commented on JetMETCorrections/Modules/plugins/QGLikelihoodESProducer.cc:
```diff
@@ -0,0 +1,58 @@
+
+// Produce the data
+std::shared_ptr<const QGLikelihoodObject> QGLikelihoodESProducer::produce(const QGLikelihoodRcd& iRecord) {
+  return std::shared_ptr<const QGLikelihoodObject>(&iRecord.get(token_), edm::do_nothing_deleter());
```
Is the point of this class and QGLikelihoodSystematicsESProducer to act as an alias for the `QGLikelihood(Systematics)Object`? (i.e. provide a constant product label for the clients, while being able to have a single point where to make changes in case the product needs to point to something else in the DB(?))

perrotta replied:
> Is the point of this class and QGLikelihoodSystematicsESProducer to act as an alias for the `QGLikelihood(Systematics)Object`? (i.e. provide a constant product label for the clients, while being able to have a single point where to make changes in case the product needs to point to something else in the DB(?))

@rappoccio may know the answer

fabiocos replied:
I would move forward with the integration of this PR, @santocch please have a look, leaving the clarification by @rappoccio as useful for possible further iterations

makortel commented:
@cmsbuild, please abort

cmsbuild commented:
Jenkins tests are aborted.

makortel committed:
Remove setIntervalFor() as unnecessary from QGLikelihood(Systematics)ESProducer

The function is not called by the framework (because these classes do
not inherit from edm::EventSetupRecordIntervalFinder), so it's clearer
to remove them.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27216/10388

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #27216 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 731f7b5c7f9ce3d541fe5ab7d811290824432356
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-094fad/966/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-094fad/966/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3254096
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3253761
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

santocch commented:
+1

perrotta commented on JetMETCorrections/Modules/interface/JetCorrectionESChain.h:
```diff
@@ -1,33 +1,5 @@
-private:
-  std::vector<std::string> mCorrectors;
-};
 #endif
```
Why not simply removing the whole file?

makortel replied:
That was my intention, but apparently I forgot to actually do that. Thanks for pointing it out.

makortel committed:
Really remove JetCorrectionESChain.h

cmsbuild commented:
The code-checks are being triggered in jenkins.

slava77 commented:
code-checks

the last attempt apparently got stuck

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27216/10415

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27216/10415/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27216/10415/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

makortel committed:
Code format

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27216/10424

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #27216 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 6eab31aa2fda698dc140af937959e80d49438e48
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-094fad/1007/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-094fad/1007/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3254096
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3253761
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

perrotta commented:
+1
- Kind of technical, following the directions outlined in the PR description
- Jenkins tests pass and show no difference

fabiocos commented:
+1

fabiocos commented:
merge

santocch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
